[quant] Add torchbind support for embedding_bag packed weights

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42881 [quant] Add torchbind support for embedding_bag packed weights**
* #42762 [quant] Add embeddingbag_prepack function that works on quantized tensor.

Summary:
This enables serialization/de-serialization of embedding packed params using getstate/setstate calls.
Added version number to deal with changes to serialization formats in future.

This can be extended in the future to support 4-bit/2-bit once we add support for that.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070634](https://our.internmc.facebook.com/intern/diff/D23070634)

supriyar committed:
[quant] Add torchbind support for embedding_bag packed weights
Summary:
This enables serialization/de-serialization of embedding packed params using getstate/setstate calls.
Added version number to deal with changes to serialization formats in future.

This can be extended in the future to support 4-bit/2-bit once we add support for that.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/embedding_packed_params.h:
```diff
@@ -0,0 +1,18 @@
+#include <ATen/ATen.h>
+#include <ATen/core/ivalue.h>
+
+struct EmbeddingPackedParamsBase : public torch::jit::CustomClassHolder {
```
nit: suggest to use `EmbeddingBagPackedParamsBase` to be more clear

supriyar replied:
I think this struct can be shared by embedding and embedding_bag operators since it only does weight packing.

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -385,10 +386,65 @@ torch::class_<LinearPackedParamsBase> register_linear_params() {
+            std::vector<double> doubles_to_serialize = {};
+            int bit_rate = params->bit_rate();
+            int version = params->version();
+            std::vector<int64_t> longs_to_serialize = {bit_rate, version};
```
hmm, should we have a separate field for version(probably as a first field as well)? in case the data format changes depending on version, we can have an easy access to version and decide what to do next

supriyar replied:
Sounds good.

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -385,10 +386,65 @@ torch::class_<LinearPackedParamsBase> register_linear_params() {
+torch::class_<EmbeddingPackedParamsBase> register_embedding_params() {
+  // Type for __getstate__/__setstate__ serialization
+  //
+  // Element 0 is a string to say what kind of EmbeddingParam this is
```
also I feel it might be better to have an int(enum) to indicate the type, since this will be serialized for each instance of embedding bag packed params

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag.cpp:
```diff
@@ -291,10 +389,39 @@ Tensor embedding_bag_4bit_rowwise_offsets(
+ public:
+  static at::Tensor run(
+      const c10::intrusive_ptr<EmbeddingPackedParamsBase>& packed_weight,
+      const Tensor& indices,
+      const Tensor& offsets,
+      const bool /* scale_grad_by_freq */,
+      const int64_t /* mode */,
+      bool sparse,
+      const c10::optional<Tensor>& per_sample_weights_,
+      const c10::optional<Tensor>& compressed_indices_mapping,
+      bool include_last_offset) {
```
how do we decide whether these arguments should be in `run` or `prepack`?

jerryzh168 replied:
asking because conv have some arguments(strde, padding etc.) included in prepack API(https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/conv.py#L201) instead of run, I recall it's because we need these parameters when packing the weights. I guess it's clearer to keep them in `run` in this case.

supriyar replied:
Yeah, in this case all these arguments aren't needed for packing, so might get confusing if we use them in prepack.

supriyar committed:
Update on "[quant] Add torchbind support for embedding_bag packed weights"


Summary:
This enables serialization/de-serialization of embedding packed params using getstate/setstate calls.
Added version number to deal with changes to serialization formats in future.

This can be extended in the future to support 4-bit/2-bit once we add support for that.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -385,10 +387,72 @@ torch::class_<LinearPackedParamsBase> register_linear_params() {
+            int64_t version = params->version();
+            std::vector<int64_t> longs_to_serialize = {bit_rate};
+            return EmbeddingParamsSerializationType(
+              static_cast<int64_t>(EmbeddingPackedParamType::QEMBEDDING_BAG),
```
actually, what are the different types? maybe we can also encode this in the type of the type of the embeddingpackedparams itself? e.g. we have Conv2dPackedParams, Conv1dPackedParams. not suggesting we should do that, but I think we should consider both options.

supriyar replied:
The types will be dependent on the tensor dtype, currently we have 8-bit and 4-bit (planned). From the op side I was planning to keep it same (eg. `quantized::embedding_bag_prepack`), since we can look at tensor dtype and decide if we should do 8-bit or 4-bit packing. 
Since we have one op, maybe having same name for the PackedParams might make more sense. Let me know if you feel that we should encode it in the name and also have different ops for 8-bit vs 4-bit (does it make a difference for graph mode?). 
The actual quantized op will be different i.e. quantized::embedding_bag_byte vs quantized::embedding_bag_4bit

jerryzh168 replied:
Are you talking about `bit_rate`? I'm talking about the EmbeddingPackedParamType here actually. Do you have a doc describing all the variations we plan to support?

jerryzh168 reviewed (approved):
Looks good, had one question about the type

supriyar committed:
Update on "[quant] Add torchbind support for embedding_bag packed weights"


Summary:
This enables serialization/de-serialization of embedding packed params using getstate/setstate calls.
Added version number to deal with changes to serialization formats in future.

This can be extended in the future to support 4-bit/2-bit once we add support for that.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070634](https://our.internmc.facebook.com/intern/diff/D23070634)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add torchbind support for embedding_bag packed weights"


Summary:
This enables serialization/de-serialization of embedding packed params using getstate/setstate calls.
Added version number to deal with changes to serialization formats in future.

This can be extended in the future to support 4-bit/2-bit once we add support for that.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070634](https://our.internmc.facebook.com/intern/diff/D23070634)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add torchbind support for embedding_bag packed weights"


Summary:
This enables serialization/de-serialization of embedding packed params using getstate/setstate calls.
Added version number to deal with changes to serialization formats in future.

This can be extended in the future to support 4-bit/2-bit once we add support for that.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070634](https://our.internmc.facebook.com/intern/diff/D23070634)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add torchbind support for embedding_bag packed weights"


Summary:
This enables serialization/de-serialization of embedding packed params using getstate/setstate calls.
Added version number to deal with changes to serialization formats in future.

This can be extended in the future to support 4-bit/2-bit once we add support for that.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070634](https://our.internmc.facebook.com/intern/diff/D23070634)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add torchbind support for embedding_bag packed weights"


Summary:
This enables serialization/de-serialization of embedding packed params using getstate/setstate calls.
Added version number to deal with changes to serialization formats in future.

This can be extended in the future to support 4-bit/2-bit once we add support for that.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070634](https://our.internmc.facebook.com/intern/diff/D23070634)

[ghstack-poisoned]
